### PR TITLE
Add DirectCommitWorkStream for direct path

### PR DESCRIPTION
**Pull request by m-trieu**
Add a DirectCommitWorkStream that has a self contained commit queue, and commit thread.
Callers will call `queueCommit` on the stream, and and commits will be batched and sent to windmill internally.

R: @scwhittle 
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcDirectCommitWorkStream.java
Comment (scwhittle) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcDirectCommitWorkStream.java:140]: instead of ComputationState how about Consumer bound to computationState by caller?
Instead of just failedCommit could have commitCompleteConsumer taking a commit and Windmill.CommitStatus.

Seems like it would be easier to test in isolation.

Comment (m-trieu) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcDirectCommitWorkStream.java:140]: Added another consumer for onCommitComplete vs reusing failed commits

I was a bit unsure of what it would look like to use a Consumer bound to computation state by the caller, can you clarify? Thanks!

Comment (scwhittle) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcDirectCommitWorkStream.java:145]: I meant that this class could avoid having to know about ComputationState, that the 
`state.completeWorkAndScheduleNextWorkForKey(`
calls could live within the failedConsumer or commitCompleteConsumer.

ie something like this:
```
Consumer<CompletedCommit> consumer = (CompletedCommit c) -> {  state.completeWorkAndScheduleNextWorkForKey(....); }
```

Looks like you might have to pass in the computationId with the commit.

**Comment by m-trieu** [comment]
@scwhittle ready for another look thanks!

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcDirectCommitWorkStream.java
Comment (scwhittle) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcDirectCommitWorkStream.java:55]: could we use composition instead of inheritence? it seems like functionality is just added on top of the existing stream and could just be a separate object containing a stream it delegates to.

https://en.wikipedia.org/wiki/Composition_over_inheritance#:~:text=Composition%20over%20inheritance%20(or%20composite,from%20a%20base%20or%20parent

It seems like https://github.com/apache/beam/pull/30312 is ending up there, with a committer object that is wrapping the stream. Perhaps we should just get that one in first and then integrate that object for direct path.

**Review by scwhittle — CHANGES_REQUESTED** [review]

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcDirectCommitWorkStream.java
Comment (m-trieu) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcDirectCommitWorkStream.java:55]: sgtm

Comment (m-trieu) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcDirectCommitWorkStream.java:55]: will incorporate https://github.com/apache/beam/pull/30312 once that pr is merged

**Comment by lostluck** [comment]
R: @scwhittle

**Pull request closed without merging**
